File: NugetRestore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;
using System.Xml.Linq;

class NuGetRestorer
{
    private string projectFolder;
    private string outputFolder;
    private string artifactoryUrl;
    private string nugetPackagesFolder;
    private HashSet<(string, string)> downloadedPackages;
    private object lockObject = new object();
    private int maxWorkers;

    public NuGetRestorer(string projectFolder, string outputFolder, string artifactoryUrl, int maxWorkers = 5)
    {
        this.projectFolder = projectFolder;
        this.outputFolder = outputFolder;
        this.artifactoryUrl = artifactoryUrl;
        this.nugetPackagesFolder = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".nuget/packages");
        this.downloadedPackages = new HashSet<(string, string)>();
        this.maxWorkers = maxWorkers;

        if (!Directory.Exists(this.nugetPackagesFolder))
        {
            Directory.CreateDirectory(this.nugetPackagesFolder);
        }
    }

    public List<PackageInfo> ReadAllPackages()
    {
        var allPackages = new List<PackageInfo>();

        foreach (var projectFile in Directory.GetFiles(projectFolder, "*.csproj"))
        {
            var packages = ReadPackageReferences(projectFile);
            allPackages.AddRange(packages);
        }

        return allPackages;
    }

    public List<PackageInfo> ReadPackageReferences(string projectFile)
    {
        var packages = new List<PackageInfo>();

        try
        {
            XDocument doc = XDocument.Load(projectFile);
            XNamespace ns = "http://schemas.microsoft.com/developer/msbuild/2003";

            foreach (var packageReference in doc.Descendants(ns + "PackageReference"))
            {
                var packageName = packageReference.Attribute("Include")?.Value;
                var packageVersion = packageReference.Attribute("Version")?.Value;
                packages.Add(new PackageInfo { Name = packageName, Version = packageVersion });
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error parsing {projectFile}: {ex.Message}");
        }

        return packages;
    }

    public async Task<string> DownloadPackageAsync(PackageInfo package)
    {
        var packageUrl = $"{artifactoryUrl}/{package.Name}/{package.Version}/{package.Name}.{package.Version}.nupkg";

        using (HttpClient client = new HttpClient())
        {
            var response = await client.GetAsync(packageUrl);

            if (response.IsSuccessStatusCode)
            {
                var packageFilePath = Path.Combine(outputFolder, $"{package.Name}.{package.Version}.nupkg");

                using (var fileStream = File.Create(packageFilePath))
                {
                    await response.Content.CopyToAsync(fileStream);
                }

                return packageFilePath;
            }
            else
            {
                Console.WriteLine($"Failed to download package {package.Name} version {package.Version}. Status code: {response.StatusCode}");
                return null;
            }
        }
    }

    public void RestorePackage(PackageInfo package)
    {
        lock (lockObject)
        {
            if (!downloadedPackages.Contains((package.Name, package.Version)))
            {
                // Download the package
                var packageFilePath = DownloadPackageAsync(package).Result;

                if (!string.IsNullOrEmpty(packageFilePath))
                {
                    // Extract the package to the NuGet packages folder
                    using (var zip = System.IO.Compression.ZipFile.OpenRead(packageFilePath))
                    {
                        zip.ExtractToDirectory(nugetPackagesFolder, true);
                    }

                    // Cleanup: Remove the downloaded package file
                    File.Delete(packageFilePath);

                    // Update the set of downloaded packages
                    downloadedPackages.Add((package.Name, package.Version));
                }
            }
            else
            {
                Console.WriteLine($"Package {package.Name} version {package.Version} is already downloaded.");
            }
        }
    }

    public void RestorePackages()
    {
        var allPackages = ReadAllPackages();

        Parallel.ForEach(allPackages, new ParallelOptions { MaxDegreeOfParallelism = maxWorkers }, package =>
        {
            RestorePackage(package);
        });
    }
}

class PackageInfo
{
    public string Name { get; set; }
    public string Version { get; set; }
}

class Program
{
    static void Main(string[] args)
    {
        if (args.Length < 3)
        {
            Console.WriteLine("Usage: dotnet run projectFolder=<project_folder> outputFolder=<output_folder> artifactoryUrl=<artifactory_url> [maxWorkers=<max_workers>]");
            Environment.Exit(1);
        }

        Dictionary<string, string> arguments = args
            .Select(arg => arg.Split('='))
            .ToDictionary(arg => arg[0].ToLower(), arg => arg.Length > 1 ? arg[1] : "");

        string projectFolder = arguments.GetValueOrDefault("projectfolder", "");
        string outputFolder = arguments.GetValueOrDefault("outputfolder", "");
        string artifactoryUrl = arguments.GetValueOrDefault("artifactoryurl", "");
        int maxWorkers = int.Parse(arguments.GetValueOrDefault("maxworkers", "5"));

        if (string.IsNullOrEmpty(projectFolder) || string.IsNullOrEmpty(outputFolder) || string.IsNullOrEmpty(artifactoryUrl))
        {
            Console.WriteLine("Missing required arguments. Usage: dotnet run projectFolder=<project_folder> outputFolder=<output_folder> artifactoryUrl=<artifactory_url> [maxWorkers=<max_workers>]");
            Environment.Exit(1);
        }

        NuGetRestorer nugetRestorer = new NuGetRestorer(projectFolder, outputFolder, artifactoryUrl, maxWorkers);
        nugetRestorer.RestorePackages();
    }
}

File: NuGetRestore.py
import os
import sys
import xml.etree.ElementTree as ET
import requests
from zipfile import ZipFile

class NuGetRestorer:
    def __init__(self, project_folder, output_folder, artifactory_url):
        self.project_folder = project_folder
        self.output_folder = output_folder
        self.artifactory_url = artifactory_url
        self.nuget_packages_folder = os.path.expanduser('~/.nuget/packages')
        self.downloaded_packages = set()

        if not os.path.exists(self.nuget_packages_folder):
            os.makedirs(self.nuget_packages_folder)

    def read_package_references(self, project_file):
        packages = []
        try:
            tree = ET.parse(project_file)
            root = tree.getroot()
            ns = {'ns': 'http://schemas.microsoft.com/developer/msbuild/2003'}

            for item_group in root.findall('.//ns:ItemGroup', ns):
                for package_reference in item_group.findall('ns:PackageReference', ns):
                    package_name = package_reference.get('Include')
                    package_version = package_reference.get('Version')
                    packages.append({'name': package_name, 'version': package_version})
        except ET.ParseError as e:
            print(f"Error parsing {project_file}: {e}")

        return packages

    def download_package(self, package_name, package_version):
        url = f'{self.artifactory_url}/{package_name}/{package_version}/{package_name}.{package_version}.nupkg'
        response = requests.get(url)

        if response.status_code == 200:
            package_file_path = os.path.join(self.output_folder, f'{package_name}.{package_version}.nupkg')
            with open(package_file_path, 'wb') as package_file:
                package_file.write(response.content)
            return package_file_path
        else:
            print(f"Failed to download package {package_name} version {package_version}. Status code: {response.status_code}")
            return None

    def restore_packages(self):
        for project_file in os.listdir(self.project_folder):
            if project_file.endswith('.csproj'):
                project_path = os.path.join(self.project_folder, project_file)
                packages = self.read_package_references(project_path)

                for package in packages:
                    package_name = package['name']
                    package_version = package['version']

                    if (package_name, package_version) not in self.downloaded_packages:
                        # Download the package
                        package_file_path = self.download_package(package_name, package_version)

                        if package_file_path:
                            # Extract the package to the NuGet packages folder
                            with ZipFile(package_file_path, 'r') as zip_ref:
                                zip_ref.extractall(self.nuget_packages_folder)

                            # Cleanup: Remove the downloaded package file
                            os.remove(package_file_path)

                            # Update the set of downloaded packages
                            self.downloaded_packages.add((package_name, package_version))
                    else:
                        print(f"Package {package_name} version {package_version} is already downloaded.")

if __name__ == "__main__":
    if len(sys.argv) != 4:
        print("Usage: python script.py <project_folder> <output_folder> <artifactory_url>")
        sys.exit(1)

    project_folder = sys.argv[1]
    output_folder = sys.argv[2]
    artifactory_url = sys.argv[3]

    nuget_restorer = NuGetRestorer(project_folder, output_folder, artifactory_url)
    nuget_restorer.restore_packages()
